File: 总题库/33.二叉搜索树的后序遍历序列.py
class Solution:
    def verifyPostorder(self, postorder: List[int]) -> bool:
        def recursive(postorder: List[int], left: int, right: int) -> bool:
            if right - left < 0:
                return True

            rootValue = postorder[right]
            index = left
            while index < right and postorder[index] < rootValue:
                index += 1

            for i in range(index, right):
                if postorder[i] < rootValue:
                    return False

            return recursive(postorder, left, index - 1) and recursive(postorder, index, right - 1)

        return recursive(postorder, 0, len(postorder) - 1)

File: 总题库/33.二叉搜索树的后序遍历序列.cpp
class Solution {
public:
    bool verifyPostorder(vector<int>& postorder) {
        return verifyPostorder(postorder, 0, postorder.size() - 1);
    }

    bool verifyPostorder(vector<int>& postorder, int left, int right) {
        if (right - left < 0) return true;
        int root = postorder.back();
        int index = left;
        while (index < right && postorder[index] < root) index++;
        for (int i = index; i < right; i++) {
            if (postorder[i] < root) return false;
        }
        return verifyPostorder(postorder, left, index - 1) && verifyPostorder(postorder, index, right - 1);
    }
};